lib/OpenQA/Schema/ResultSet/Assets.pm

```diff
@@ -57,48 +52,44 @@ sub register ($self, $type, $name, $options = {}) {
         });
 }
 
-sub scan_for_untracked_assets {
-    my ($self) = @_;
+sub scan_for_untracked_assets ($self) {
 
     # search for new assets and register them
+    my $assetdir = assetdir();
     for my $type (TYPES) {
-        my @paths;
 
-        my $assetdir = assetdir();
+        my @paths;
         for my $subtype (qw(/ /fixed)) {
             my $path = "$assetdir/$type$subtype";
-            my $dh;
-            next unless opendir($dh, $path);
+            next unless opendir(my $dh, $path);
             for my $file (readdir($dh)) {
-                unless ($file eq 'fixed' or $file eq '.' or $file eq '..') {
-                    push(@paths, "$path/$file");
-                }
+                next if $file eq 'fixed' || $file eq '.' || $file eq '..';
+                push(@paths, "$path/$file");
             }
             closedir($dh);
         }
 
-        my %paths;
+        my %known = map { $_->name => 1 } $self->search({type => $type})->all;
+        my @register;
         for my $path (@paths) {
+            my $name = basename($path);
+            next if $known{$name};
 
-            my $basepath = basename($path);
             # ignore links
             next if -l $path;
 
             # ignore files not owned by us
-            next unless -o $path;
+            next unless -o _;
```

Comment:
I assume `_` refers to the last internal `stat`. Maybe it would make sense to add an explaining comment nevertheless because the syntax is quite "magic".

It could also be just one line:
```
# ignore links, files not owned by us and non-existing files and folders
# note: `_` refers to the result of the last internal `stat`.
next if -l $path || !-o _ || !-e _;
```

Reply:
Using the `_` filehandle is not really magic. It's pretty standard Perl for reusing the result of the previous stat call for `-...` operators and explained in `perldoc -f -e`.

Reply:
I mean not self-explaining and not easy to google for people who don't know Perl very well.

Reply:
The idea of explaining Perl features that are in the Perl documentation with comments bugs me a little. Which other Perl features would get the same treatment? Where would we draw the line? Let's hear what others on the team think.

Reply:
Maybe reference that part of the documentation without repeating the content? Is it https://perldoc.perl.org/functions/stat that mentions "If stat is passed the special filehandle consisting of an underline, no stat is done, but the current contents of the stat structure from the last stat, lstat, or filetest are returned"?

I like the suggestion of @Martchus to use

```perl
# ignore links, files not owned by us and non-existing files and folders
# note: `_` refers to the result of the last internal `stat`.
next if -l $path || !-o _ || !-e _;
```

Reply:
See https://perldoc.perl.org/functions/-e

Reply:
I think better is:
```
# ignore links, files not owned by us and non-existing files and folders
next if  -l -e -o $path;
```

Reply:
> I think better is:
> 
> ```perl
> # ignore links, files not owned by us and non-existing files and folders
> next if  -l -e -o $path;
> ```

The logic is wrong.

Reply:
Anyway, i think i've addressed all  concerns in the latest revision.